api/handlers.go

```diff
@@ -92,22 +110,30 @@ func (api *API) loginGET(w http.ResponseWriter, req *http.Request, _ httprouter.
 
 // loginPOST starts a user session by issuing a cookie
 func (api *API) loginPOST(w http.ResponseWriter, req *http.Request, _ httprouter.Params) {
+	defer api.closeBody(req)
+	// Get the body, we might need to use it several times.
+	body, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		api.WriteError(w, errors.AddContext(err, "empty request body"), http.StatusBadRequest)
+		return
+	}
+
 	// Since we don't want to have separate endpoints for logging in with
 	// credentials and token, we'll do both here.
 	//
 	// Check whether credentials are provided. Those trump the token because a
 	// user with a valid token might want to relog. No need to force them to
 	// log out first.
-	email := req.PostFormValue("email")
-	pw := req.PostFormValue("password")
-	if email != "" && pw != "" {
-		api.loginPOSTCredentials(w, req, email, pw)
+	var payload credentialsDTO
+	err = json.Unmarshal(body, &payload)
+	if err == nil && payload.Email != "" && payload.Password != "" {
+		api.loginPOSTCredentials(w, req, payload.Email, payload.Password)
 		return
 	}
 
 	// Check for a challenge response in the request.
 	var chr database.ChallengeResponse
-	err := chr.LoadFromRequest(req)
+	err = chr.LoadFromRequest(bytes.NewBuffer(body))
```

Comment:
why is it called `LoadFromRequest` if you pass it a reader?

Reply:
Because I forgot to rename it. :) It used to read from the request.
I'll fix it, thanks.